File: mpm/python/usrp_mpm/xports/xport_adapter_ctrl.py
#
# Copyright 2022 Ettus Research, a National Instruments Brand
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
"""
Transport adapter control
"""

from enum import Enum
import netaddr
from usrp_mpm.mpmutils import poll_with_timeout
from usrp_mpm.compat_num import CompatNumber
from usrp_mpm.mpmlog import get_logger
from usrp_mpm.sys_utils.uio import UIO

MIN_COMPAT_NUM_REMOTE_STREAMING = CompatNumber(1, 0)

# pylint: disable=too-many-arguments

class XportAdapterCtrl:
    """
    Controls advanced transport adapter (TA) capabilities, if present.

    Note that this requires the ability to read/write from the transport adapter
    address space, which is not always available (e.g., on older bitfiles).
    Make sure to not instantiate one of these classes if the feature is not
    available, or it will result in a bus error.

    This class will use the compat number of the TA to determine its
    capabilities, which it will store in the 'features' attribute. Valid entries
    for this attribute include:
    - rx_routing: This TA has the capability to route RX data to arbitrary endpoints
    - rx_hdr_removal: The TA can remove CHDR headers when routing data to
                      arbitrary endpoints (StreamModes.RAW_PAYLOAD may be used)
    """
    # Address offsets (these must match xport_sv/eth_regs.vh)
    # pylint: disable=bad-whitespace
    XPORT_ADAPTER_COMPAT_NUM      = 0x0100 # 8 bits major, 8 bits minor
    XPORT_ADAPTER_INFO            = 0x0104
    XPORT_ADAPTER_NODE_INST       = 0x0108 # read-only
    KV_MAC_LO                     = 0x010C
    KV_MAC_HI                     = 0x0110
    KV_IPV4                       = 0x0114
    KV_UDP_PORT                   = 0x0118
    KV_CFG                        = 0x011C
    # pylint: enable=bad-whitespace

    class StreamModes(Enum):
        """
        Valid streaming modes for outgoing packets

        The numerical values must be such that they can be written to KV_CFG
        (see eth_regs.vh).
        """
        FULL_PACKET = 0
        RAW_PAYLOAD = 1


    def __init__(self, label):
        self.log = get_logger(label)
        self._regs = UIO(label=label, read_only=False)
        self.poke32 = self._regs.poke32
        self.peek32 = self._regs.peek32
        self.features = set()
        self._ta_index = -1
        # These three registers are always available, even if the transport
        # adapter supports nothing else.
        with self._regs:
            compat_num = self.peek32(self.XPORT_ADAPTER_COMPAT_NUM)
            self._ta_index = self.peek32(self.XPORT_ADAPTER_NODE_INST)
            adapter_info = self.peek32(self.XPORT_ADAPTER_INFO)
        self._compat_num = CompatNumber((compat_num >> 8) & 0xFF, compat_num & 0xFF)
        # If the FPGA is new enough to know about remote UDP streaming, but
        # doesn't have that feature enabled, it will have a compat number of
        # zero.
        if self._compat_num < MIN_COMPAT_NUM_REMOTE_STREAMING:
            return
        if bool(adapter_info & (1 << 0)):
            self.features.add('rx_routing')
        if bool(adapter_info & (1 << 1)):
            self.features.add('rx_hdr_removal')

    def get_xport_adapter_inst(self):
        """
        Returns an integer value that tells us which transport adapter is
        connected to this interface (the instance, or node_inst) value.
        """
        return self._ta_index

    def get_compat_num(self):
        """
        Return the compat number of this transport adapter
        """
        return self._compat_num

    def add_remote_ep_route(self, epid, ipv4, port, mac_addr, stream_mode):
        """
        Adds a route from this transport adapter to a CHDR streaming endpoint
        with endpoint ID epid.
        When CHDR packets with the given epid reach the transport adapter, they
        will be sent to the ipv4/port/mac_addr destination.

        :param epid: The 16-bit endpoint ID value.
        :param ipv4: A string representation of the destination IPv4 address
        :param port: An integer representation of the destination port
        :param mac_addr: A string representation of the destination MAC address
        :param stream_mode: The stream mode used for outgoing packets
        """
        # Sanitize inputs
        assert isinstance(stream_mode, self.StreamModes)
        assert (epid >> 16) == 0
        # Check capabilities
        if 'rx_routing' not in self.features:
            raise RuntimeError(
                "This transport adapter does not support routing to remote "
                "destinations!")
        if stream_mode == self.StreamModes.RAW_PAYLOAD \
                and 'rx_hdr_removal' not in self.features:
            raise RuntimeError(
                "Requesting to remove CHDR headers, but feature not enabled!")
        # Encode inputs for our registers
        stream_mode_int = stream_mode.value
        mac_addr_int = int(netaddr.EUI(mac_addr))
        mac_addr_lo = mac_addr_int & 0xFFFFFFFF
        mac_addr_hi = (mac_addr_int >> 32) & 0xFFFF
        ipv4_int = int(netaddr.IPAddress(ipv4))
        port = int(port)
        cfg_word = epid | (stream_mode_int << 16)
        self.log.debug(
            f"On transport adapter {self._ta_index}: Adding route from EPID "
            f"{epid} to destination {ipv4}:{port} (MAC Address: {mac_addr}), "
            f"stream mode {stream_mode.name} ({stream_mode_int})")
        # Now write registers
        with self._regs:
            # Check BUSY bit before writing new values
            if not poll_with_timeout(
                    lambda: not bool(self.peek32(self.KV_CFG) & (1 << 31)),
                    0.5, # timeout at 500 ms
                    0.1, # check at 100 ms interval
                ):
                raise RuntimeError(
                    f"Timeout while polling BUSY bit on transport adapter "
                    f"{self._ta_index}!")
            # Now write all the configurations; CFG goes last
            self.poke32(self.KV_MAC_LO, mac_addr_lo)
            self.poke32(self.KV_MAC_HI, mac_addr_hi)
            self.poke32(self.KV_IPV4, ipv4_int)
            self.poke32(self.KV_UDP_PORT, port)
            self.poke32(self.KV_CFG, cfg_word)

File: host/lib/usrp/cores/xport_adapter_ctrl.cpp
//
// Copyright 2022 Ettus Research, a National Instruments Brand
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <uhd/exception.hpp>
#include <uhd/utils/log.hpp>
#include <uhdlib/usrp/cores/xport_adapter_ctrl.hpp>
#include <unordered_map>
#include <boost/algorithm/string/case_conv.hpp>
#include <boost/asio.hpp>
#include <chrono>
#include <cstdio>
#include <thread>

using namespace uhd::usrp;

namespace {

const uhd::compat_num16 MIN_COMPAT_REMOTE_STRM{1, 0};

const std::unordered_map<std::string, uint32_t> STREAM_MODES{
    {xport_adapter_ctrl::STREAM_MODE_FULL_PACKET, 0},
    {xport_adapter_ctrl::STREAM_MODE_RAW_PAYLOAD, 1}};

std::pair<uint32_t, uint32_t> cast_ipv4_and_port(
    const std::string& ipv4, const std::string& port)
{
    using namespace boost::asio;
    io_service io_service;
    ip::udp::resolver resolver(io_service);
    try {
        ip::udp::resolver::query query(ip::udp::v4(), ipv4, port);
        ip::udp::endpoint endpoint = *resolver.resolve(query);
        return {
            uint32_t(endpoint.address().to_v4().to_ulong()), uint32_t(endpoint.port())};
    } catch (const std::exception&) {
        throw uhd::value_error("Invalid UDP address: " + ipv4 + ":" + port);
    }
}

std::pair<uint32_t, uint32_t> cast_mac(const std::string& mac_addr)
{
    unsigned char mac[8] = {0}; // 8 bytes for conversion to uint64_t
    int ret              = std::sscanf(mac_addr.c_str(),
        "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
        &mac[5],
        &mac[4],
        &mac[3],
        &mac[2],
        &mac[1],
        &mac[0]);
    if (ret != 6) {
        throw uhd::value_error("Invalid MAC address: " + mac_addr);
    }
    const uint32_t mac_msb = *(reinterpret_cast<uint32_t*>(mac + 0));
    const uint32_t mac_lsb = *(reinterpret_cast<uint32_t*>(mac + 4));
    return {mac_msb, mac_lsb};
}


} // namespace

constexpr const char xport_adapter_ctrl::STREAM_MODE_RAW_PAYLOAD[] = "RAW_PAYLOAD";
constexpr const char xport_adapter_ctrl::STREAM_MODE_FULL_PACKET[] = "FULL_PACKET";

xport_adapter_ctrl::xport_adapter_ctrl(poke_fn_type&& poke_fn,
    peek_fn_type&& peek_fn,
    const bool has_arp,
    const std::string& log_id)
    : _poke32(std::move(poke_fn))
    , _peek32(std::move(peek_fn))
    , _log_id(log_id)
    , _compat_num(_peek32(XPORT_ADAPTER_COMPAT_NUM))
    , _ta_inst(
          _compat_num >= MIN_COMPAT_REMOTE_STRM ? _peek32(XPORT_ADAPTER_NODE_INST) : 0)
{
    if (_compat_num >= MIN_COMPAT_REMOTE_STRM) {
        const uint32_t capabilities = _peek32(XPORT_ADAPTER_INFO);
        if (capabilities & (1 << 0)) {
            _capabilities["rx_routing"] = "1";
        }
        if (capabilities & (1 << 1)) {
            _capabilities["rx_hdr_removal"] = "1";
        }
        if (has_arp) {
            _capabilities["arp"] = "1";
        }
        _capabilities["ta_inst"] = std::to_string(_ta_inst);
    }
    UHD_LOG_TRACE(
        _log_id, "Remote streaming capabilities: " << _capabilities.to_string());
}


void xport_adapter_ctrl::add_remote_ep_route(const uhd::rfnoc::sep_inst_t epid,
    const std::string ipv4,
    const std::string port,
    const std::string mac_addr,
    const std::string stream_mode)
{
    const std::string stream_mode_upper = boost::to_upper_copy(stream_mode);
    UHD_ASSERT_THROW(STREAM_MODES.count(stream_mode_upper));
    if (!_capabilities.has_key("rx_routing")) {
        throw uhd::runtime_error(
            "This transport adapter does not support routing to remote "
            "destinations!");
    }
    if (stream_mode_upper == STREAM_MODE_RAW_PAYLOAD
        && !_capabilities.has_key("rx_hdr_removal")) {
        throw uhd::runtime_error(
            "Requesting to remove CHDR headers, but feature not enabled!");
    }

    if (ipv4.empty()) {
        throw uhd::value_error("Must provide valid IPv4 address!");
    }
    if (port.empty()) {
        throw uhd::value_error("Must provide valid port value!");
    }
    if (mac_addr.empty() && !_capabilities.has_key("arp")) {
        throw uhd::value_error(
            "Device has no ARP capabilities -- must provide MAC address!");
    }
    const bool request_arp         = mac_addr.empty();
    const auto ipv4_and_port       = cast_ipv4_and_port(ipv4, port);
    const uint32_t stream_mode_int = STREAM_MODES.at(stream_mode_upper);
    const uint32_t cfg_word        = epid | (stream_mode_int << 16);

    // Check TA is ready by polling BUSY flag
    using namespace std::chrono_literals;
    const auto timeout = std::chrono::steady_clock::now() + 500ms;
    while (bool(_peek32(KV_CFG) & (1 << 31))) {
        if (std::chrono::steady_clock::now() > timeout) {
            UHD_LOG_THROW(uhd::runtime_error,
                _log_id,
                "Timeout while polling BUSY flag on transport adapter!");
        }
        std::this_thread::sleep_for(100ms);
    }

    // Now write settings to TA
    UHD_LOG_DEBUG(_log_id,
        "On transport adapter " << _ta_inst << ": Adding route from EPID " << epid
                                << " to destination " << ipv4 << ":" << port
                                << " (MAC Address: " << (request_arp ? "AUTO" : mac_addr)
                                << "), stream mode " << stream_mode_upper << " ("
                                << stream_mode_int << ")");
    if (!request_arp) {
        const auto mac_int = cast_mac(mac_addr);
        _poke32(KV_MAC_LO, mac_int.first);
        _poke32(KV_MAC_HI, mac_int.second);
        _poke32(KV_IPV4, ipv4_and_port.first);
    } else {
        // If the user didn't specify MAC, then the device firmware can try and
        // look it up.
        constexpr int num_arp_tries   = 3;
        constexpr auto retry_interval = 300ms;
        bool arp_successful           = false;
        for (int i = 0; i < num_arp_tries; i++) {
            try {
                _poke32(KV_IPV4_W_ARP, ipv4_and_port.first);
                arp_successful = true;
                break;
            } catch (const uhd::lookup_error&) {
                UHD_LOG_TRACE(_log_id, "ARP lookup failed for IP address " << ipv4);
                std::this_thread::sleep_for(retry_interval);
            }
        }
        if (!arp_successful) {
            UHD_LOG_THROW(uhd::lookup_error,
                _log_id,
                "Device was unable to look up Ethernet (MAC) address for IP "
                "address "
                    << ipv4
                    << ". Make sure device is correctly connected, "
                       "or provide MAC address manually.");
        }
    }
    _poke32(KV_UDP_PORT, ipv4_and_port.second);
    _poke32(KV_CFG, cfg_word);
}
